mlflow/system_metrics/system_metrics_monitor.py

```diff
@@ -0,0 +1,94 @@
+"""Class for monitoring system stats."""
+
+import threading
+import logging
+
+from mlflow.system_metrics.metrics.cpu_monitor import CPUMonitor
+from mlflow.system_metrics.metrics.disk_monitor import DiskMonitor
+from mlflow.system_metrics.metrics.network_monitor import NetworkMonitor
+from mlflow.system_metrics.metrics.gpu_monitor import GPUMonitor
+
+from mlflow.utils.autologging_utils import BatchMetricsLogger
+
+logger = logging.getLogger(__name__)
+
+
+class SystemMetricsMonitor:
+    """Class for monitoring system stats.
+
+    This class is used for pulling system metrics and logging them to MLflow. Calling `start()` will
+    spawn a thread that logs system metrics periodically. Calling `finish()` will stop the thread.
+
+    Args:
+        mlflow_run: an 'mlflow.entities.run.Run' instance, which is used to bootstrap system metrics
+            logging with the MLflow tracking server.
+        logging_interval: float, the interval (in seconds) at which to log system metrics to MLflow.
+    """
+
+    def __init__(self, mlflow_run, logging_interval=5.0):
+        # Instantiate default monitors.
+        self.monitors = [CPUMonitor(), DiskMonitor(), NetworkMonitor()]
+        gpu_monitor = GPUMonitor()
+        if gpu_monitor:
+            self.monitors.append(gpu_monitor)
+        self.logging_interval = logging_interval
+
+        self.mlflow_logger = BatchMetricsLogger(mlflow_run.info.run_id)
+        self._shutdown_event = threading.Event()
+        self._process = None
+        # Attach `SystemMetricsMonitor` instance to the `mlflow_run` instance.
+        mlflow_run.system_metrics_monitor = self
+
+    def start(self):
+        """Start monitoring system metrics."""
+        try:
+            self._process = threading.Thread(
+                target=self.monitor,
+                daemon=True,
+                name="SystemMetricsMonitor",
+            )
+            self._process.start()
+            logger.info(f"MLflow: started monitoring system metrics.")
+        except Exception as e:
+            logger.warning(f"MLflow: failed to start monitoring system metrics: {e}")
+            self._process = None
+
+    def monitor(self):
+        """Main loop for the thread, which consistently collect and log system metrics."""
+        while not self._shutdown_event.is_set():
+            metrics = self.collect_metrics()
+            self._shutdown_event.wait(self.logging_interval)
+            if self._shutdown_event.is_set():
+                break
+            try:
+                self.log_metrics(metrics)
```

Comment:
2 things you might want to be aware of that can help with this:
1. the autologging client: https://github.com/mlflow/mlflow/blob/master/mlflow/utils/autologging_utils/client.py#L89C7-L89C38
2. async metric logging: https://docs.google.com/document/d/1uI_6AkgL_GIi_i9GP_nVSV26RhDLAIrr/edit

in general I think you want to decouple the metric computation daemon from the logging attempt (it will let you summarize/agg at a different rate from the logging attempt to avoid throttling)

As a side note, platform implementations (like ours in AzureML) have different channels to report these metrics, so I would split this into 3 parts:
1. metric collector daemon (useful for OSS)
2. async metric logging (use one of the above 2 approaches)
3. defining a system metric type/canonical names independent of the above 2 (then AzureML/others can report them in the read path without using the write paths)

Reply:
As another interesting tidbit, many DS @ microsoft who wrote one of these on their own for MLflow always forgot to guard the code for rank 0, resulting in heavy throttling as they scaled out the number of nodes 😄  something to consider

Reply:
Hi! What you are suggesting is very correct. We will have a different frequency for metrics pulling and logging, and metrics will be aggregated before sending to logging. Async logging API is in our roadmap, we will plug it onto the system metrics logging once it is available. This PR is in progress. 

I don't understand what you mean by "guard the code for rank 0", could you share more details on it? thanks!

Reply:
Sure, the rank terminology comes from MPI jobs, but the problem applies to any distributed job where the same code traditionally runs on every node. An example of something you almost never want is to log loss metrics from every rank, because after an all-reduce, logging the loss from the head/driver node/main process (usually called "rank 0") is the same and is 1/N the network traffic.

Your case is a bit more interesting because it makes a lot of sense to collect the performance metrics from all the nodes to debug data loading pipeline stalls or bad GPU nodes. However, the scaling of network traffic still holds (and your metric daemon is probably more chatty than the train batch loop). To avoid this, there are 3 options:

1. aggregate your metrics onto the driver/rank0 and log from there always, so there is no linear scaling
1.1. or only log from one rank as a "representative sample" metric rather than for diagnostics
2. delegate the problem entirely to the hardware provider when perf counters are natively available (the approach AML takes)
2.1. But even for this, a portion of your PR is still useful in the form of defining "standard metric names" and a means to get and search them on runs (the read path, basically, and no reuse of the write path)
3. reduce your traffic and its priority as much as possible
3.1. This is where using async / queued-with-major-backoff|delay logging can dramatically help as you no longer need to provide strong consistency on these metrics

Hope that helps clarify and wonder what you think!